File: assets/js/vida.js
function Vida(source, target){
	window.vidaClient = this;
	window.words = "";
	this.routes = {
	    'api:authorize':'/api/authorize',
	    'api:help#languages':'/api/help/languages',
	    'api:engine#transliterate':'/api/engine/transliterate'
	};
	this.setup = function(){
		NProgress.start();
		$.post(window.vidaClient.routes['api:authorize'], {}, function(ajax){
			if (ajax.status)
				window.csrf_token = ajax.content.csrf_token;
			else {
				console.error("Couldn't authenticate the SDK.");
				return false;
			}
			NProgress.done();
		}).fail(function(e, s, t){
			NProgress.done();
			console.error(e);
			console.error(s);
			console.error(t);
		});
		if ($.inArray(this.source, Object.keys(this.codes)) == -1){
			console.error("Please specify the correct code for the source langauge;");
			return false;
		}
		if ($.inArray(this.target, Object.keys(this.codes)) == -1){
			console.error("Please specify the correct code for the target langauge;");
			return false;
		}
	};
	this.codes = languages = {
		"hin": "Hindi",
		"ben": "Bengali",
		"guj": "Gujarati",
		"pun": "Punjabi",
		"mal": "Malayalam",
		"kan": "Kannada",
		"tam": "Tamil",
		"tel": "Telugu",
		"ori": "Oriya",
		"mar": "Marathi",
		"ass": "Assamese",
		"kon": "Konkani",
		"bod": "Bodo",
		"nep": "Nepali",
		"urd": "Urdu",
		"eng": "English"
	};
	this.source = source;
	this.target = target;
	this.text = null;
	this.config = function(name, value){
		if (name == 'source')
			this.source = value;
		if (name == 'target')
			this.target = value;
	};
	this.setTransliteratedText = function(id, word, replacement){
		$(id).val($(id).val().replace(new RegExp(word, 'gi'), replacement));
	};
	this.startEngine = function(id){
		this.setup();
		$(id).keypress(function(e){
			if (e.keyCode == 32){
				var text = $(id).val().split(" ");
				var word = text[text.length -1];
				$.post(window.vidaClient.routes['api:engine#transliterate'], {
					csrf_token:window.csrf_token,
					source:this.source,
					target:this.target,
					text:word
				}, function(ajax){
					if (ajax.status) 
						window.vidaClient.setTransliteratedText(id, word, ajax.content.output);
				}).fail(function(e, s, t){
					console.log(e, s, t)
				});
			};
		});
	};
};

File: libraries/vida.py
from repo import Repo
from indictrans import Transliterator
import enchant


class Vida:
	"""
	Only focused on English to 15 Indic Language Transliteration
	Hindi (hin)
	Bengali (ben)
	Gujarati (guj)
	Punjabi (pun)
	Malayalam (mal)
	Kannada (kan)
	Tamil (tam)
	Telugu (tel)
	Oriya (ori)
	Marathi (mar)
	Assamese (ass)
	Konkani (kon)
	Bodo (bod)
	Nepali (nep)
	Urdu (urd)
	English (eng)
	"""

	languages = {
		"hin": "Hindi",
		"ben": "Bengali",
		"guj": "Gujarati",
		"pun": "Punjabi",
		"mal": "Malayalam",
		"kan": "Kannada",
		"tam": "Tamil",
		"tel": "Telugu",
		"ori": "Oriya",
		"mar": "Marathi",
		"ass": "Assamese",
		"kon": "Konkani",
		"bod": "Bodo",
		"nep": "Nepali",
		"urd": "Urdu",
		"eng": "English"
	}

	codes = languages.keys()

	@staticmethod
	def is_ascii(s): return all(ord(c) < 128 for c in s)

	def __init__(self, text, source, target):
		self.text = text
		self.source = source
		self.target = target
		self.isEngSource = (self.source == 'eng')
		if self.isEngSource:
			self.usdictionary = enchant.Dict("en_US")
			self.gbdictionary = enchant.Dict("en_GB")
		self.validated = self.source in self.codes and self.target in self.codes
		if self.validated:
			self.engine = Transliterator(source=self.source, target=self.target)
		
	def run(self):
		status = False; message = "Couldn't transliterate the text."; content = {}
		output = []
		if not self.validated:
			message = "Please provide languages and their code."
			output = self.text
		else:
			text = self.text.split()
			try:
				for index in xrange(len(text)):
					word = text[index]
					if not self.isEngSource:
						word = word.decode('utf-8')
						output.insert(index, self.engine.transform(word).encode('utf-8'))
					else:
						if not Vida.is_ascii(word): word = word.decode('utf-8')
						if not self.usdictionary.check(word) and not self.gbdictionary.check(word):
							output.insert(index, self.engine.transform(word).encode('utf-8'))
						else: 
							output.insert(index, word)
				status = True
				message = "Succesfully transliterated the code."
			except UnicodeDecodeError, e:
				Repo.exception(e)
				message = "Couldn't decode the language properly."
			except IndexError, e:
				Repo.exception(e)
				message = "Couldn't properly frame the sentence."
			output = ' '.join(output)
		output = output.decode('utf-8')
		content['output'] = output
		return Repo.api('libraries:vida#run', status, message, content)